blocks/library/image/block.js

```diff
@@ -257,4 +243,16 @@ class ImageBlock extends Component {
 	}
 }
 
-export default withEditorSettings()( ImageBlock );
+export default flowRight( [
+	withEditorSettings(),
+	withAPIData( ( props ) => {
+		const { id } = props.attributes;
+		if ( ! id ) {
+			return {};
+		}
+
+		return {
+			image: `/wp/v2/media/${ id }`,
```

Comment:
It's surprising that you need to combine the endpoint's url here, but otherwise it's very nice 👍

Reply:
>It's surprising that you need to combine the endpoint's url here

Could you elaborate on what you mean by this? I'm not sure if the surprise is something I should be worried about 😄

Reply:
I would expect to see API url creation hidden in the HOC. So in this place we could have the following instead:
```js
return [
	id && { type: 'image', params: { id } }
];
```
`withAPIData` could filter out undefined elements to make this more conceise here.

At the moment we have the following code in a few places:
```js
const applyWithAPIData = withAPIData( () => {
	return {
		user: '/wp/v2/users/me?context=edit',
	};
} );```

which would simplify to:
```js
const applyWithAPIData = withAPIData( () => [ { type: 'user' } ] );
```

`withAPIData` would have all mappings stored internally:
`user` => `/wp/v2/users/me?context=edit`

I might be missing something obvious here, but I'm happy to discuss if you think it would be an improvement. It's good how it is because those urls won't rather change in the future :)

Reply:
The issue I'd see is `type` wouldn't always be so simple. This could potentially be used for custom endpoints, which can be defined as whichever route structure the developer chooses (even `/this/is/an/obnoxious/route/pattern`). I noted some downsides here in [the original text](https://gist.github.com/aduth/e5dd6327963475e9f580b8ece523ffa0) of #1974, including how this binds us to the REST API, but doing so in a way that is easy for developers to use (don't need to understand mappings if they already know the routes), isolated, generic.